### PR TITLE
Add Config docroot detection, fixes #663

### LINKED ISSUE #663: "Docroot Location" should pick up normal directory roots
## What happened (or feature request): 
Feature request

## What you expected to happen:
When I configured a composer based Drupal project, I figured it would pick up `web` or `docroot` when "Docroot Location"  was asked.

## How to reproduce this:
Run `ddev config`

## Related source links or issues:

Example of what I did in my tool `platform-docker`

```
                // Platform.sh scaffold docroot.
                if (is_dir($this->cwd . '/' . Platform::DEFAULT_WEB_ROOT)) {
                  Config::set('docroot', Platform::DEFAULT_WEB_ROOT);
                }
                // Typical app location.
                elseif (is_dir($this->cwd . '/web')) {
                    Config::set('docroot', 'web');
                }
                // Acquia.
                elseif (is_dir($this->cwd . '/docroot')) {
                    Config::set('docroot', 'docroot');
                }
                else {
                  $question = new Question("<info>What is the document root for the project?");
                  $answer = $helper->ask($input, $output, $question);
                  Config::set('docroot', $answer);
                }
```

source: https://github.com/mglaman/platform-docker/blob/develop/src/Command/InitCommand.php#L48

## Please use a complexity rating of 1-5 (5 is high) for a feature request. (High complexity implies more PR planning)

1

### PR DESCRIPTION
**Pull request by mglaman**
## The Problem/Issue/Bug:

OP #663 - Our users would love to have docroot detected for them during `ddev config`.

## How this PR Solves The Problem:

Check if any of the following directories exist and default to the first one found

* web
* docroot
* htdocs
* _www (legacy Platform.sh)
* public

## Manual Testing Instructions:

Run `ddev config` and see that the docroot discovery re-uses `app.Docroot` or detects an existing directory if present.

## Automated Testing Overview:

Added `TestConfigCommandDocrootDetection` which tests automatic detection of `web`, `htdocs`, and `docroot`.

## Related Issue Link(s):

#663

### CODE REVIEW

**Comment by mglaman** [comment]
Almost.

```
Creating a new ddev project config in the current directory (/tmp/TestConfigCommand_web204449359)
Once completed, your configuration will be written to /tmp/TestConfigCommand_web204449359/.ddev/config.yaml

Project name (): 
The docroot is the directory from which your site is served. This is a relative path from your application root (/tmp/TestConfigCommand_web204449359)
You may leave this value blank if your site files are in the application root
Docroot Location: Found a php codebase at /tmp/TestConfigCommand_web204449359/web.
Application Type [drupal7, drupal8, wordpress, typo3, backdrop, php, drupal6] (php): Creating a new ddev project config in the current directory (/tmp/TestConfigCommand_docroot392530274)
Once completed, your configuration will be written to /tmp/TestConfigCommand_docroot392530274/.ddev/config.yaml

Project name (): 
The docroot is the directory from which your site is served. This is a relative path from your application root (/tmp/TestConfigCommand_docroot392530274)
You may leave this value blank if your site files are in the application root
Docroot Location: Found a php codebase at /tmp/TestConfigCommand_docroot392530274/docroot.
Application Type [backdrop, php, drupal6, drupal7, drupal8, wordpress, typo3] (php): Creating a new ddev project config in the current directory (/tmp/TestConfigCommand_htdocs481897305)
Once completed, your configuration will be written to /tmp/TestConfigCommand_htdocs481897305/.ddev/config.yaml

Project name (): 
The docroot is the directory from which your site is served. This is a relative path from your application root (/tmp/TestConfigCommand_htdocs481897305)
You may leave this value blank if your site files are in the application root
Docroot Location: Found a php codebase at /tmp/TestConfigCommand_htdocs481897305/htdocs.
```

It's not showing the default in the prompt, but test is passing!

**Comment by mglaman** [comment]
Symfony uses `public` https://symfony.com/doc/current/setup/web_server_configuration.html. It looks like Laravel uses the same. Same with Slim PHP framework.

That should probably be added.

**Comment by mglaman** [comment]
I added _www and public. I wasn't sure how to make it a shared definition so we didn't duplicate in config and config_test. I'm still pro Goland n00b

**Comment by rfay** [comment]
Awesome! Thanks for this. Looking forward to reviewing it.

**Comment by rickmanelius** [comment]
Updated reviewers and put on our project board.

File: pkg/ddevapp/config.go [hunk 1]
```diff
@@ -365,12 +365,23 @@ func (app *DdevApp) docrootPrompt() error {
 	output.UserOut.Printf("\nThe docroot is the directory from which your site is served. This is a relative path from your application root (%s)", app.AppRoot)
 	output.UserOut.Println("You may leave this value blank if your site files are in the application root")
 	var docrootPrompt = "Docroot Location"
-	if app.Docroot != "" {
-		docrootPrompt = fmt.Sprintf("%s (%s)", docrootPrompt, app.Docroot)
+	// Provide use the app.Docroot as the default docroot option.
+	var defaultDocroot = app.Docroot
+	if defaultDocroot == "" {
+		for _, docroot := range []string{"web", "docroot", "htdocs", "_www", "public"} {
```
Comment (rfay) [inline pkg/ddevapp/config.go]: Let's define these outside of inline code please. They should be the same in tests as in the definition code. Just a global array definition should be fine.

Comment (rfay) [reply pkg/ddevapp/config.go]: Please make this work with web/public (normal composer location for TYPO3) as well.

Comment (rfay) [reply pkg/ddevapp/config.go]: Shouldn't it be looking in the approot/project root first? Or am I missing something here?

File: pkg/ddevapp/config_test.go [hunk 2]
```diff
@@ -194,6 +194,56 @@ func TestConfigCommand(t *testing.T) {
 	}
 }
 
+// TestConfigCommandDocrootDetection asserts the default docroot is detected.
+func TestConfigCommandDocrootDetection(t *testing.T) {
+	// Set up tests and give ourselves a working directory.
+	assert := asrt.New(t)
+
+	testMatrix := []string{"web", "docroot", "htdocs", "_www", "public"}
```
Comment (rfay) [inline pkg/ddevapp/config_test.go]: As above, these should be defined outside of the code so both test and application can use the same values in the same order.

File: pkg/ddevapp/config_test.go [hunk 3]
```diff
@@ -194,6 +194,56 @@ func TestConfigCommand(t *testing.T) {
...
+	for _, testDocrootName := range testMatrix {
+		testDir := testcommon.CreateTmpDir("TestConfigCommand_" + testDocrootName)
+
+		// testcommon.Chdir()() and CleanupDir() checks their own errors (and exit)
+		defer testcommon.CleanupDir(testDir)
+		defer testcommon.Chdir(testDir)()
+
+		// Create a document root folder.
+		err := os.Mkdir(filepath.Join(testDir, testDocrootName), 0644)
```
Comment (rfay) [inline pkg/ddevapp/config_test.go]: Use os.MkdirAll() here, for n-level docroots.

File: pkg/ddevapp/config.go (hunk 1 above)
Comment (mglaman) [reply pkg/ddevapp/config.go]: Perhaps we should....

* web/public (TYPO3, otherwise default composer template wins)
* web
* docroot
* htdocs
* public
* _www
* _default to approot_

Comment (rfay) [reply pkg/ddevapp/config.go]: Probably if approot is already defined we have to stick with it. So it should be the default? But then we have the ambiguity of "" being the docroot vs "" being unset.

Comment (mglaman) [reply pkg/ddevapp/config.go]: I guess I meant default being "" which is approot? or is approot == docroot?

All I meant was "Default the last option to being the current directory." Which maybe hitting enter does that" Maybe just say `Docroot location: (current directory)`

Comment (rfay) [reply pkg/ddevapp/config.go]: Approot or project root is the base of the repo (or where .ddev has been created). Docroot is where index.php is.  Sometimes they are the same.

Comment (mglaman) [reply pkg/ddevapp/config.go]: Cool. So I think we cycle through the list. If nothing provided as default the prompt should be `Docroot location: (current directory)` so you know you can hit enter and go. That's the current behavior, but without the hint, correct?

Comment (rfay) [reply pkg/ddevapp/config.go]: 👍

**Comment by mglaman** [comment]
💩 `web/public` broke `testDir := testcommon.CreateTmpDir("TestConfigCommand_" + testDocrootName)` in test because of dirname.

**Comment by rfay** [comment]
Please go ahead and fix that in this PR with filepath.Join(). It's pretty silly.

**Comment by mglaman** [comment]
There seems to be test errors with the TYPO3 directory I'm going to investigate while looking into the `---sitename` issue as well.

File: cmd/ddev/cmd/config.go [hunk 4]
```diff
@@ -107,7 +107,10 @@ var ConfigCommand = &cobra.Command{
 				if _, err = os.Stat(docrootRelPath); os.IsNotExist(err) {
 					util.Failed("The docroot provided (%v) does not exist", docrootRelPath)
 				}
+			} else if !cmd.Flags().Changed("docroot") {
+				app.Docroot = ddevapp.DiscoverDefaultDocroot(app)
```
Comment (mglaman) [inline cmd/ddev/cmd/config.go]: @rfay This is my workaround. If the `docroot` was not changed, default to detected directory.

Comment (rfay) [reply cmd/ddev/cmd/config.go]: Looks correct to me.

File: cmd/ddev/cmd/config_test.go [hunk 5]
```diff
@@ -40,3 +42,31 @@ func TestConfigDescribeLocation(t *testing.T) {
 	assert.Contains(string(out), "No project configuration currently exists")
 
 }
+
```
Comment (rfay) [inline cmd/ddev/cmd/config_test.go]: Nit: Add comment explaining the test.

**Comment by rickmanelius** [comment]
I ran through my tests again and this indeed works. I do like the addition of the check for an index.php in the trial docroot directory. I do think it might be important to check the current directory for an index.php file to negate searching one level deeper, but that's probably a rare edge case. And if it does come up, we can tweak.

@rfay it looks like a comment as added to describe the test. I believe that was your only remaining request. If so, I'm good with merging if you are!

File: pkg/ddevapp/config.go [hunk 6]
```diff
@@ -354,6 +354,36 @@ func (app *DdevApp) promptForName() error {
...
+// DiscoverDefaultDocroot returns the default docroot directory.
+func DiscoverDefaultDocroot(app *DdevApp) string {
+	// Provide use the app.Docroot as the default docroot option.
+	var defaultDocroot = app.Docroot
+	if defaultDocroot == "" {
+		for _, docroot := range AvailableDocrootLocations() {
+			if _, err := os.Stat(docroot); err != nil {
+				continue
+			}
+			if fileutil.FileExists(docroot + "/index.php") {
```
Comment (rfay) [inline pkg/ddevapp/config.go]: This needs to be filepath.Join() instead of string concat. Good old Windows!

**Comment by rfay** [comment]
Oh and please rebase in the process so it gets all caught up, thanks!

**Comment by mglaman** [comment]
Rebased. Test should create an `index.php` to fix the with flags test.

**Comment by rfay** [comment]
So awesome @mglaman - Thanks for this excellent work. Big step forward for ddev. It's needed this for a while.
